features/cellular/framework/device/CellularStateMachine.h

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2018, Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef _CELLULAR_STATEMACHINE_H_
+#define _CELLULAR_STATEMACHINE_H_
+
+#include "EventQueue.h"
+#include "CellularNetwork.h"
+#include "CellularCommon.h"
+#include "Semaphore.h"
+
+namespace rtos {
+    class Thread;
+}
+
+namespace mbed {
+
+class FileHandle;
+class UARTSerial;
+class CellularPower;
+class CellularSIM;
+class CellularDevice;
+
+const int RETRY_ARRAY_SIZE = 10;
+
+/** CellularStateMachine class
+ *
+ *  Finite State Machine for connecting to cellular network.
+ *  By default automatic reconnecting is on. This means that when FSM gets the disconnected callback
+ *  it will try to connect automatically. Application can toggle this behavior with method set_automatic_reconnect(...)
+ */
+class CellularStateMachine {
+public:
+    /** Constructor
+     *
+     * @param device    reference to CellularDevice
+     * @param queue     reference to queue used in state transitions
+     * @param power     power needed in first state. Can be also given with set_power but must be given before
+     *                  calling run_to_state. Transfers ownership to this class.
+     */
+    CellularStateMachine(CellularDevice &device, events::EventQueue &queue, FileHandle &fh);
+    ~CellularStateMachine();
+
+public:
+    /** Cellular connection states
```

Comment:
In general we should try to make everything private, otherwise users use these classes instead of APIs.